### PR TITLE
[#9230][refactor] Replace nemotron patches with custom model implementation

### LINKED ISSUE #9230: [Feature]: Add a slimmed down reference implementation for nemotron H models
### 🚀 The feature, motivation and pitch

The reference nemotron H code on HuggingFace is out of date, and therefore bugged, and has several untested code paths. This makes an already hairy patching system even hairier.

The proposal is to do away with those patches, and replace the entire `AutoModelForCausalLM` implementation with one that is heavily slimmed down. Ideally, it should:

* work (fix bugs in the current implementation)
* add no new dependencies to TensorRT-LLM
* do away with unnecessary features for TensorRT-LLM / AutoDeploy:
  - no training related code (dropout, gradient checkpointing, etc.)
  - no caching business (we want to replace it with our own anyway)
  - no attention masking where possible
  - ...
* reuse existing AD custom ops for mamba SSM update / causal conv1d

If successful, this will also show an alternative path to the current patching approach.

### Alternatives

- The current patching approach.

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Summary by CodeRabbit

* **New Features**
  * Added support for Nemotron H model architecture with optimized inference components.

* **Improvements**
  * Enhanced FP8 quantization configuration for better pattern matching and validation accuracy.
  * Refined shape propagation logic for quantization operations.

* **Refactor**
  * Streamlined model implementation by replacing legacy patching infrastructure with direct component implementations.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## Description

* Why?

Patching for nemotron H models was growing out of hand, and made certain
optimizations more complex than they needed to be.

* What?

This commit finally gets rid of them, and replaces them with the custom
model implementation in `modeling_nemotron_h.py`.

Closes #9230
Closes NvBug 5747867

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/transform/library/quantization.py [hunk 1]
```diff
@@ -123,7 +123,7 @@ def _apply(
             cnt += 1
 
         return gm, TransformInfo(
-            skipped=False, num_matches=cnt, is_clean=False, has_valid_shapes=True
+            skipped=False, num_matches=cnt, is_clean=False, has_valid_shapes=(cnt == 0)
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/auto_deploy/transform/library/quantization.py]: Actually `has_valid_shapes` should probably just be hardcoded to `False`.

File: tensorrt_llm/_torch/auto_deploy/models/patches/nemotron_h.py
[the remark this answers is not in the document]
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/models/patches/nemotron_h.py]: where is this being used such that it causes a circular dependency? How about we place it in `tensorrt_llm/_torch/auto_deploy/custom_ops/rms_norm.py`? Shouldn't cause any issues there

**Review by lucaslie — APPROVED** [review]
Please do a final rebase to main and graph/perf check. Thank you so much for pushing this along

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 2]
```diff
@@ -317,8 +281,43 @@ def __init__(self, config, layer_idx: Optional[int] = None):
             layer_idx=layer_idx,
         )
 
-    # TODO: inline code from `_nemotron_h_moe_forward` when removing patches.
-    forward = _nemotron_h_moe_forward
+    def forward(self, hidden_states: torch.Tensor):
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: At time of writing, this is still the case: https://github.com/NVIDIA/TensorRT-LLM/blob/7175d89b483c459cf7fbd55cb7fcfb2c99184ee8/tensorrt_llm/_torch/auto_deploy/models/patches/nemotron_h.py#L121

File: tensorrt_llm/_torch/auto_deploy/transform/library/quantization.py [hunk 3]
```diff
@@ -123,7 +123,7 @@ def _apply(
             cnt += 1
 
         return gm, TransformInfo(
-            skipped=False, num_matches=cnt, is_clean=False, has_valid_shapes=True
+            skipped=False, num_matches=cnt, is_clean=False, has_valid_shapes=False
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/transform/library/quantization.py]: Done.

File: tests/unittest/_torch/auto_deploy/unit/singlegpu/models/test_modeling_nemotron_h.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    atol = 0.05
+
+    torch.testing.assert_close(test, ref, rtol=rtol, atol=atol)
+
+
+@pytest.mark.parametrize(
+    "model_dir,model_on_meta_during_export",
+    [
+        ("nvidia/NVIDIA-Nemotron-Nano-12B-v2", True),
+        ("nvidia/NVIDIA-Nemotron-Nano-12B-v2", False),
```
Comment (lucaslie) [inline tests/unittest/_torch/auto_deploy/unit/singlegpu/models/test_modeling_nemotron_h.py]: Please make sure to use llm-models root in CI instead of pulling from HF every time. Related to this nvbug: Please make sure to address the nvbug: https://nvbugspro.nvidia.com/bug/5747867

File: tensorrt_llm/_torch/auto_deploy/models/custom/modeling_nemotron_h.py
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/models/custom/modeling_nemotron_h.py]: Done.
